frontend/src/app/workspace/service/copilot/texera-copilot.ts

```diff
@@ -0,0 +1,282 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Injectable } from "@angular/core";
+import { BehaviorSubject, Observable, from, of, throwError, defer } from "rxjs";
+import { map, catchError, tap, switchMap, finalize } from "rxjs/operators";
+import { WorkflowActionService } from "../workflow-graph/model/workflow-action.service";
+import {
+  toolWithTimeout,
+  createGetOperatorInCurrentWorkflowTool,
+  createGetOperatorPropertiesSchemaTool,
+  createGetOperatorPortsInfoTool,
+  createGetOperatorMetadataTool,
+  createListAllOperatorTypesTool,
+  createListLinksInCurrentWorkflowTool,
+  createListOperatorsInCurrentWorkflowTool,
+} from "./workflow-tools";
+import { OperatorMetadataService } from "../operator-metadata/operator-metadata.service";
+import { createOpenAI } from "@ai-sdk/openai";
+import { generateText, type ModelMessage, stepCountIs } from "ai";
+import { WorkflowUtilService } from "../workflow-graph/util/workflow-util.service";
+import { AppSettings } from "../../../common/app-setting";
+import { WorkflowCompilingService } from "../compile-workflow/workflow-compiling.service";
+import { COPILOT_SYSTEM_PROMPT } from "./copilot-prompts";
+import { NotificationService } from "../../../common/service/notification/notification.service";
+
+export enum CopilotState {
+  UNAVAILABLE = "Unavailable",
+  AVAILABLE = "Available",
+  GENERATING = "Generating",
+  STOPPING = "Stopping",
+}
+
+export interface AgentUIMessage {
+  role: "user" | "agent";
+  content: string;
+  isBegin: boolean;
+  isEnd: boolean;
+  toolCalls?: any[];
+  toolResults?: any[];
+  usage?: {
+    inputTokens?: number;
+    outputTokens?: number;
+    totalTokens?: number;
+    cachedInputTokens?: number;
+  };
+}
+
+/**
+ * Texera Copilot Service provides AI-powered assistance for workflow creation and manipulation.
+ *
+ * This service manages a single AI agent instance that can:
+ * 1. Interact with users through natural language messages
+ * 2. Execute workflow operations using specialized tools
+ * 3. Maintain conversation history and state
+ *
+ * The service communicates with an LLM backend (via LiteLLM) to generate responses and uses
+ * workflow tools to perform actions like listing operators, getting operator schemas, and
+ * manipulating workflow components.
+ *
+ * State management includes:
+ * - UNAVAILABLE: Agent not initialized
+ * - AVAILABLE: Agent ready to receive messages
+ * - GENERATING: Agent currently processing and generating response
+ * - STOPPING: Agent in the process of stopping generation
+ */
+@Injectable()
+export class TexeraCopilot {
+  private model: any;
+  private modelType = "";
+  private agentName = "";
+  private messages: ModelMessage[] = [];
```

Comment:
Can you add some comments about these variables? Both `messages` and `agentResponses` are messages and it is not immediately clear which does what.